src/FileCache.js

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) 2017 Electric Imp
+// This file is licensed under the MIT License
+// http://opensource.org/licenses/MIT
+
+'use strict';
+
+const fs = require('fs-extra');
+const url = require('url');
+const path = require('path');
+const minimatch = require('minimatch');
+const HttpReader = require('./Readers/HttpReader');
+const GithubReader = require('./Readers/GithubReader');
+const GITHUB_DIR = path.sep + 'github';
+const HTTP_DIR = path.sep + 'http';
+const DEFAULT_EXCLUDE_FILE_NAME = 'builder-cache.exclude';
+const CACHED_READERS = [GithubReader, HttpReader];
+const CACHE_LIFETIME = 1; // in days
+
+class FileCache {
+
+  constructor(machine) {
+    this._useCache = false;
+    this._cacheDir = '.' + path.sep + '.builder-cache';
+    this._excludeList = [];
+    this._machine = machine;
+    this._outdateTime = CACHE_LIFETIME * 86400000; // precalc milliseconds in one day
+  }
+
+  /**
+   * Transform url or github link to path and filename
+   * It is important, that path and filename are unique,
+   * because collision can break the build
+   * @param {string} link link to the file
+   * @return {string} folder and name, where cache file can be found
+   * @private
+   */
+  _getCachedPath(link) {
+    const ghRes = GithubReader.parseUrl(link);
+    if (ghRes !== false) {
+      return this._getCachedGithubPath(ghRes);
+    }
+    if (new HttpReader().supports(link)) {
+      return this._getCachedHttpPath(link);
+    }
+  }
+
+  /**
+   * Transform github link to path and filename
+   * @param {string} link link to the file
+   * @return {string} folder and name, where cache file can be found
+   * @private
+   */
+  _getCachedHttpPath(httpLink) {
+    // parse url parts
+    const parsedUrl = url.parse(httpLink);
+    const domain = parsedUrl.hostname.split('.'); // it is web-site name
+    // create new path from url
+    const newPath = this.cacheDir + HTTP_DIR + path.join.apply(domain.filter((elem) => elem != 'www').reverse())
+      + (parsedUrl.pathname ? parsedUrl.pathname.replace(/\//g, path.sep) : '');
+    return newPath;
+  }
+
+  /**
+   * Transform url link to path and filename
+   * @param {{user, repo, path, ref}} ghRes github parsed link to the file
+   * @return {string} folder and name, where cache file can be found
+   * @private
+   */
+  _getCachedGithubPath(ghRes) {
+    // find, where fileName starts
+    const i = ghRes.path.lastIndexOf('/');
+    let newPath;
+    let fileName;
+    // check case, when filename goes after user and repo
+    if (i != -1) {
+      newPath = ghRes.path.substr(0, i).replace(/\//g, path.sep);
+      fileName = ghRes.path.substr(i + 1);
+    }
+    const filePath = this.cacheDir + GITHUB_DIR + path.sep + ghRes.user + path.sep + ghRes.repo
```

Comment:
probably yes, cose some one can include one file from 2 forked repos, this files may be different but we can save only one of them (because of name's collision)